### PR TITLE
feat: listen for forum events

### PR DESCRIPTION
**Pull request by Ian2012**
### Description

This PR adds listeners for the forum events created here: https://github.com/openedx/openedx-events/pull/273 and emitted here https://github.com/openedx/edx-platform/pull/33374/

### Testings instructions:
1. Install `tutor-forum` plugin
2. Enable the plugin and run `tutor config save`
3. Build the tutor forum image: `tutor images build forum`
4. Install `openedx-events` and `edx-platform` from the branches mentioned above
5. Init the environment
6. Import the demo course
7. Go to the discussions page
8. Create a new post
9. An email notification should be sent. Those can be found in the lms container in the folder: `tmp/openedx/emails`

**Merge checklist:**
Check off if complete *or* not applicable:
- [ ] Version bumped
- [ ] Changelog record added
- [ ] Documentation updated (not only docstrings)
- [ ] Fixup commits are squashed away
- [ ] Unit tests added/updated
- [ ] Manual testing instructions provided
- [ ] Noted any: Concerns, dependencies, migration issues, deadlines, tickets

### CODE REVIEW

File: platform_plugin_forum_email_notifier/email.py [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+from openedx.core.lib.celery.task_utils import emulate_http_request
+
+
+class ForumEmailNotification(MessageType):
+    """
+    Message type for forum email notification.
+    """
+
+
+forum_email_message_type = ForumEmailNotification()
```
Comment (mariajgrimaldi) [inline platform_plugin_forum_email_notifier/email.py]: Why initialize this here?

File: platform_plugin_forum_email_notifier/apps.py [hunk 2]
```diff
@@ -11,3 +11,23 @@ class PlatformPluginForumEmailNotifierConfig(AppConfig):
...
+                "test": {"relative_path": "settings.test"},
+                "production": {"relative_path": "settings.production"},
+            },
+        },
+    }
+
+    def ready(self):
+        super().ready()
+
+        from platform_plugin_forum_email_notifier import email, tasks, handlers  # pylint: disable=unused-import, import-outside-toplevel
```
Comment (mariajgrimaldi) [inline platform_plugin_forum_email_notifier/apps.py]: why are we registering the email module? is it necessary?

File: platform_plugin_forum_email_notifier/email.py [hunk 3]
```diff
@@ -0,0 +1,27 @@
+"""Email module for platform_plugin_forum_email_notifier"""
+from django.contrib.sites.models import Site
+from edx_ace import ace
+from edx_ace.message import MessageType
+from openedx.core.lib.celery.task_utils import emulate_http_request
```
Comment (mariajgrimaldi) [inline platform_plugin_forum_email_notifier/email.py]: can we wrap this in a try-except?

File: platform_plugin_forum_email_notifier/email.py [hunk 4]
```diff
@@ -0,0 +1,27 @@
+"""Email module for platform_plugin_forum_email_notifier"""
```
Comment (mariajgrimaldi) [inline platform_plugin_forum_email_notifier/email.py]: what does the email module do?

File: platform_plugin_forum_email_notifier/handlers.py [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+from platform_plugin_forum_email_notifier.tasks import send_email_notification
+from platform_plugin_forum_email_notifier.utils import get_subscribers
+
+
+@receiver(FORUM_THREAD_CREATED)
+def forum_thread_created_handler(signal, sender, thread, metadata, **kwargs):  # pylint: disable=unused-argument
+    """
+    Handler for forum thread created event.
+    """
+    notify_users(thread, object_type="thread")
```
Comment (mariajgrimaldi) [inline platform_plugin_forum_email_notifier/handlers.py]: object_type could be an enum so we can reuse it

File: platform_plugin_forum_email_notifier/handlers.py [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+
+@receiver(FORUM_RESPONSE_COMMENT_CREATED)
+def forum_comment_created_handler(signal, sender, thread, metadata, **kwargs):  # pylint: disable=unused-argument
+    """
+    Handler for forum comment created event.
+    """
+    notify_users(thread, object_type="comment")
+
+
+def notify_users(thread, object_type):
```
Comment (mariajgrimaldi) [inline platform_plugin_forum_email_notifier/handlers.py:51]: why isn't this a task as well? we could have a large number of subscribers

File: platform_plugin_forum_email_notifier/tasks.py [hunk 7]
```diff
@@ -0,0 +1,57 @@
+"""Tasks for the platform_plugin_forum_email_notifier plugin."""
+import logging
+
+from celery import shared_task
+from django.contrib.auth import get_user_model
+from edx_ace.recipient import Recipient
+from edx_django_utils.monitoring import set_code_owner_attribute
+from openedx.core.djangoapps.content.course_overviews.models import CourseOverview
```
Comment (mariajgrimaldi) [inline platform_plugin_forum_email_notifier/tasks.py]: There's a method called: get_course_overview_or_none in edx-platform https://github.com/openedx/edx-platform/blob/fcc80eca30ab81ec9387eb264bcf4035d175025b/openedx/core/djangoapps/content/course_overviews/api.py#L17-L32

File: platform_plugin_forum_email_notifier/templates/platform_plugin_forum_email_notifier/edx_ace/forumemailnotification/email/body.html [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+{% block content %}
+<table width="100%" align="left" border="0" cellpadding="0" cellspacing="0" role="presentation">
+    <tr>
+        <td>
+            <h1>
+                {% trans "Forum Activity" as tmsg %}{{ tmsg | force_escape }}
+            </h1>
+            <p style="color: rgba(0,0,0,.75);">
+                {% filter force_escape %}
+                {% blocktrans %}Activity in the discussion :{% endblocktrans %}
```
Comment (mariajgrimaldi) [inline platform_plugin_forum_email_notifier/templates/platform_plugin_forum_email_notifier/edx_ace/forumemailnotification/email/body.html]:
```suggestion
                {% blocktrans %}Activity in the discussion: {% endblocktrans %}
```

File: platform_plugin_forum_email_notifier/templates/platform_plugin_forum_email_notifier/edx_ace/forumemailnotification/email/body.txt [hunk 9]
```diff
@@ -0,0 +1,6 @@
+{% load i18n %}{% autoescape off %}
+{% blocktrans %}We received a request to change the e-mail associated with your {{ platform_name }} account from {{ old_email }} to {{ new_email }}. If this is correct, please confirm your new e-mail address by visiting:{% endblocktrans %}
+
+{{ confirm_link }}
+
+{% blocktrans %}If you didn't request this, you don't need to do anything; you won't receive any more email from us. Please do not reply to this e-mail; if you require assistance, check the help section of the {{ platform_name }} web site.{% endblocktrans %}{% endautoescape %}
```
Comment (mariajgrimaldi) [inline platform_plugin_forum_email_notifier/templates/platform_plugin_forum_email_notifier/edx_ace/forumemailnotification/email/body.txt:6]: what's this file for?

File: platform_plugin_forum_email_notifier/utils.py [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+
+from openedx.core.djangoapps.django_comment_common.comment_client import settings, utils
+
+Subscriber = namedtuple("Subscriber", ["id"])
+
+
+def get_subscribers(thread_id):
+    """Return a list of user ids subscribed to a thread."""
+    page = 1
+    num_pages = 100
```
Comment (mariajgrimaldi) [inline platform_plugin_forum_email_notifier/utils.py:17]: should we make this configurable? Or at least store them as constants

File: platform_plugin_forum_email_notifier/utils.py [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+            subscribers.append(Subscriber(id=subscription["subscriber_id"]))
+
+        num_pages = response["num_pages"]
+        page = response["page"]
+        page += 1
+
+    return subscribers
+
+
+def _url_for_thread_subscriptions(thread_id):
```
Comment (mariajgrimaldi) [inline platform_plugin_forum_email_notifier/utils.py:35]: what's prefix?

Comment (mariajgrimaldi) [reply platform_plugin_forum_email_notifier/utils.py:35]: Could you add docstrings to the function?

File: platform_plugin_forum_email_notifier/utils.py [hunk 12]
```diff
@@ -0,0 +1,32 @@
+"""Utilities for the platform_plugin_forum_email_notifier plugin."""
+
+from collections import namedtuple
+
+from openedx.core.djangoapps.django_comment_common.comment_client import settings, utils
```
Comment (mariajgrimaldi) [inline platform_plugin_forum_email_notifier/utils.py]: We can use try excepts here or backends.

File: platform_plugin_forum_email_notifier/utils.py (hunk 11 above)
Comment (Ian2012) [reply platform_plugin_forum_email_notifier/utils.py:35]: In tutor is `forum:4567/api/v1/`, but other deployment options may use a different prefix. It's implemented using the same pattern as the official forum client

File: platform_plugin_forum_email_notifier/tasks.py [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+from openedx.core.djangoapps.lang_pref import LANGUAGE_KEY
+from openedx.core.djangoapps.user_api.preferences.api import get_user_preference
```
Comment (mariajgrimaldi) [inline platform_plugin_forum_email_notifier/tasks.py]: We can use try excepts here or backends.

File: platform_plugin_forum_email_notifier/email.py (hunk 1 above)
Comment (Ian2012) [reply platform_plugin_forum_email_notifier/email.py]: ACE messages contain a global context that's used to fulfill basic information, this is done at startup, but we are not using it. We can move it if needed

File: platform_plugin_forum_email_notifier/apps.py (hunk 2 above)
Comment (Ian2012) [reply platform_plugin_forum_email_notifier/apps.py]: For hot reloading

File: platform_plugin_forum_email_notifier/email.py (hunk 4 above)
Comment (Ian2012) [reply platform_plugin_forum_email_notifier/email.py]: Sends ACE emails for forum subscribers

File: platform_plugin_forum_email_notifier/handlers.py (hunk 6 above)
Comment (Ian2012) [reply platform_plugin_forum_email_notifier/handlers.py:51]: Will be refactored in this PR https://github.com/eduNEXT/platform-plugin-forum-email-notifier/pull/3/files#diff-d6f8aa8c3a53ed681d96a188fd78330e8a56fafecdc0171b67c325cccb2648adR64-R101

File: platform_plugin_forum_email_notifier/templates/platform_plugin_forum_email_notifier/edx_ace/forumemailnotification/email/body.txt (hunk 9 above)
Comment (Ian2012) [reply platform_plugin_forum_email_notifier/templates/platform_plugin_forum_email_notifier/edx_ace/forumemailnotification/email/body.txt:6]: When sending emails it's import to have two versions of the email. The html one and the .txt one

File: platform_plugin_forum_email_notifier/utils.py (hunk 10 above)
Comment (Ian2012) [reply platform_plugin_forum_email_notifier/utils.py:17]: I don't think is needed, those are just initial values that will satisfy the initial while condition

File: platform_plugin_forum_email_notifier/templates/platform_plugin_forum_email_notifier/edx_ace/forumemailnotification/email/body.txt (hunk 9 above)
Comment (mariajgrimaldi) [reply platform_plugin_forum_email_notifier/templates/platform_plugin_forum_email_notifier/edx_ace/forumemailnotification/email/body.txt:6]: But why `we received a request to change the e-mail associated with your...`?

File: platform_plugin_forum_email_notifier/tasks.py [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+User = get_user_model()
+
+
+@shared_task
+@set_code_owner_attribute
+def send_email_notification(
+    subscriber, course_id, thread_body, thread_title, thread_url, thread_type
+):
+    """
+    Serialize a course and writes it to ClickHouse.
```
Comment (mariajgrimaldi) [inline platform_plugin_forum_email_notifier/tasks.py]: You might need to update this docstring

**Review by mariajgrimaldi — COMMENTED** [review]
I read the code and left a few comments. Now I'll be testing it out!

File: platform_plugin_forum_email_notifier/templates/platform_plugin_forum_email_notifier/edx_ace/forumemailnotification/email/body.txt (hunk 9 above)
Comment (Ian2012) [reply platform_plugin_forum_email_notifier/templates/platform_plugin_forum_email_notifier/edx_ace/forumemailnotification/email/body.txt:6]: That's copypasta. The email and text templates are not full yet, preferably @johnvente will improve the template once he got time or i will improve it in a following PR once the functional part is fully finished

File: platform_plugin_forum_email_notifier/tasks.py (hunk 14 above)
Comment (Ian2012) [reply platform_plugin_forum_email_notifier/tasks.py]: Done

File: platform_plugin_forum_email_notifier/utils.py (hunk 10 above)
Comment (mariajgrimaldi) [reply platform_plugin_forum_email_notifier/utils.py:17]: so we're always checking just for the first 100 pages?

**Review by mariajgrimaldi — APPROVED** [review]
This looks good. I haven't tried it, though. My discussions MFE is currently failing.

File: platform_plugin_forum_email_notifier/utils.py (hunk 10 above)
Comment (Ian2012) [reply platform_plugin_forum_email_notifier/utils.py:17]: no, this is updated here: 

https://github.com/eduNEXT/platform-plugin-forum-email-notifier/pull/2/files#diff-98cb4bd825d030cff04e0bd6ac2b3c1d8bf7df41f3ee8cfe75d29a31c02d3056R28

**Comment by mariajgrimaldi** [comment]
I tested cag/email-digest locally which uses this PR and it's working perfectly:
image

Please, let's try closing this PR and for any additional change we can do it in another one.

**Pull request merged**